src/mono/wasm/runtime/assets.ts

```diff
@@ -47,6 +47,40 @@ const skipInstantiateByAssetTypes: {
     "dotnetwasm": true,
 };
 
+export function get_skipped_icu_assets() : { [name: string]: boolean; }
+{
+    const preferredCulture = (navigator.languages && navigator.languages[0]);
+    const prefix = preferredCulture.split("-")[0];
+    if (["en", "fr", "it", "de", "es"].includes(prefix))
+    {
+        return {
+            "icudt.dat": true,
+            "icudt_EFIGS.dat": false,
+            "icudt_CJK.dat": true,
+            "icudt_no_CJK.dat": true
+        };
+    }
+    if (["zh", "ko", "ja",].includes(prefix)) {
+        return {
+            "icudt.dat": true,
+            "icudt_EFIGS.dat": true,
+            "icudt_CJK.dat": false,
+            "icudt_no_CJK.dat": true
+        };
+    }
+    return {
+        "icudt.dat": true,
+        "icudt_EFIGS.dat": true,
+        "icudt_CJK.dat": true,
+        "icudt_no_CJK.dat": false
+    };
+}
+
+export function skipIcuAssets(asset : AssetEntryInternal, skipIcuByAssetNames: { [name: string]: boolean; }) : boolean{
+    // if custom then there is only one icu file in assets and it should not be skipped
+    return !(asset.behavior == "icu" && !runtimeHelpers.config.isIcuDataCustom && skipIcuByAssetNames[asset.name]);
```

Comment:
could we just detect the fact that ICU data is custom by fact that in the assets, there is at least oneasset with behavior icu with name doesn't match well known names?
(instead of adding `isIcuDataCustom`)

Reply:
Not really, `IsDatafileCustom` = true also when we are loading the well known datafiles.
Now I can see that naming is deceiving now. `IsDatafileCustom` is true when the user provided value for <IcuFileName> in .csproj. They can provide even `icudt.dat` that is actually icu file shipped by us and I still expect the boolean to be true. It should be named `IsIcuFileDefined` I guess.

Reply:
`useBuiltInIcuFile` maybe ? But I still don't understand why we need extra flag.

Reply:
Let's look at `src\mono\wasm\build\WasmApp.targets`. Assuming we are not in invariant mode:
- If user did not provide any `IcuFileName`, we are populating `IcuFileName` on our own. In `assets.ts` we would be able to detect this situation as you said: checking if the icu names in mono-config are known.
- If user provided `IcuFileName` that is one of our built-in files, e.g. `icudt_EFIGS.dat`, we are not changing it, it's predefined. In mono-config there will be only one icu asset, so in `assets.ts` we do not have to check `skipIcuByAssetNames`.
- If user provided `IcuFileName` that is not our built-in file, e.g. `icudt_custom.dat`, we are not changing it, it's predefined but not built-in. In mono-config there will be only one icu asset, so in `assets.ts` we do not have to check `skipIcuByAssetNames`.
The workaround for using the flag would be checking the count of all icu behavior assets in mono-config. If 1: then assume it's predefined, if 5 then assume it's not. This workaround looks a bit unintuitive and without reading `WasmApp.targets` the person reading `assets.ts` would be confused. Do you think it's anyway better than adding that flag?

Reply:
> Do you think it's anyway better than adding that flag?

Yes. And description instead of the flag.

Reply:
My expectations could be described this way.
1) We will use the first file which matches both behavior==icu and locale (by known name)
2) We will use the first file which matches the behavior==icu

Reply:
Flag is removed, we are deciding based on the contents of `config-mono` if to check the app's locale or not.